nrf_rpc/include/nrf_rpc_common.h

```diff
@@ -0,0 +1,181 @@
+ */
+#define NRF_RPC_CONTAINER_OF(_ptr, _type, _field_name) \
+	((_type *)(((uint8_t *)_ptr) - offsetof(_type, _field_name)))
+
+#ifdef __GNUC__
+#define __NRF_RPC_STATIC_ASSERT(_expr, _msg) \
+	_Static_assert(_expr, _msg)
+#else
+#define __NRF_RPC_STATIC_ASSERT(_expr, _msg) \
+	extern char (*_ignore(void))
+#endif /* __GNUC__ */
+
+/**
+ * @brief Compile-time assert.
+ *
+ * @param _expr Condition that have to be met.
+ * @param _msg  Message shown when condition is not met.
+ */
+#define NRF_RPC_STATIC_ASSERT(_expr, _msg) \
+	__NRF_RPC_STATIC_ASSERT(_expr, _msg)
+
+#if defined(__ASSERT)
+#define __NRF_RPC_ASSERT(_expr) __ASSERT(_expr, "nRF RPC assertion failed")
+#else
+#include <assert.h>
+#define __NRF_RPC_ASSERT assert
+#endif
+
+/** @brief Runtime assert.
+ *
+ * @param _expr Condition that have to be met.
+ */
+#define NRF_RPC_ASSERT(_expr) \
+	__NRF_RPC_ASSERT(_expr)
+
+/** @brief Creates new automatically registered array.
+ *
+ * @param _name      Name of the array. Can be used later for
+ *                   @a NRF_RPC_AUTO_ARR_FOR.
+ * @param _array_key String array unique key that will be used to add items with
+ *                   @a NRF_RPC_AUTO_ARR_ITEM.
+ */
+#define NRF_RPC_AUTO_ARR(_name, _array_key)				       \
+	const uint8_t NRF_RPC_CONCAT(_name, _auto_arr_end) __used	       \
+	__attribute__((__section__(".nrf_rpc." _array_key ".c")));	       \
+	const uint8_t *const _name __used				       \
+	__attribute__((__section__(".nrf_rpc." _array_key ".a"))) =	       \
+		&NRF_RPC_CONCAT(_name, _auto_arr_end)
+
+/** @brief Adds new variable to the array.
+ *
+ * @param _type      Type of the variable.
+ * @param _name      Name of the variable.
+ * @param _array_key String array unique key.
+ * @param _item_key  String item key.
+ */
+#define NRF_RPC_AUTO_ARR_ITEM(_type, _name, _array_key, _item_key)	       \
+	_type _name __used						       \
+	__attribute__((__section__(".nrf_rpc." _array_key ".b." _item_key)))
+
+/** @brief Iterate over array items.
+ *
+ * This macro must be placed in `for` brackets.
+ *
+ * @param _it        Pointer variable of type `void *` for iterator.
+ * @param _var       Pointer variable that holds each item of the array.
+ * @param _array_ptr Pointer to array.
+ * @param _type      Type of items in array.
+ */
+#define NRF_RPC_AUTO_ARR_FOR(_it, _var, _array_ptr, _type)		       \
+	(_var) = (_type *)((const uint8_t *const *)(_array_ptr) + 1);	       \
+		(const uint8_t *const)_var <				       \
+			*(const uint8_t *const *)(_array_ptr);		       \
+		(_var) = (_type *)(_var) + 1, (void)_it
+
+/** @brief Get item from the array.
+ *
+ * @param _array_ptr Pointer to array.
+ * @param _index     Index of the item.
+ * @param _type      Type of items in array.
+ */
+#define NRF_RPC_AUTO_ARR_GET(_array_ptr, _index, _type)			       \
+	(((_type *)((const uint8_t *const *)(_array_ptr) + 1))[_index])
+
+/* Error codes definitions if they are not defined by the platform */
+#ifndef ENOENT
```

Comment:
Do we really want to extract those definition right now? As you suggested in below comment this entire file can be later used commonly in nrfxlib. Additionally those definition does not define new error codes. They are more like fixup if the OS does not have already UNIX like error codes.

Reply:
I think this library should use NRF_ERRNO error codes, instead of the OS error codes, that is how all the other libraries in nrfxlib works.

Reply:
There was already discussion related to error codes and I changed it to be compatible with `errno.h`. I have no problem changing it again, this time to NRF_ERRNO, but I want to know opinion from others to avoid voices like "but we agreed to something else before".

Reply:
I looked to the previous discussion related to error codes. I see that NRF_ERRNO was suggested as one of alternatives, so I suspect that no one will complain about it. I can switch to NRF_ERRNO, but I can see really big issue, which can be resolved only by merging all `nrf_errno.h` files.

All three files are incompatible for now. Codes are the same, but they have different set of errors. If someone will use nRF RPC together with other module with `nrf_errno.h` file then only one will be included causing some undefined identifier errors. I can create superset of all `nrf_errno.h` files, so if my file goes first in include directory list then everything will work correctly. If not then nRF PRC will not compile correctly.

I see that NRF_ERRNO approach is more elegant that current solution, so I will change it, but we have to take care of merging `nrf_errno.h` later to avoid above issue.